### DISCUSSION TITLE
Sbt run with javaagent

### DISCUSSION THREAD

**Discussion by fokot** [playframework/playframework#12195] [Q&A] [answered]
I want to run our app with opentelemetry java agent like ```java -javaagent:path/to/opentelemetry-javaagent.jar -jar myapp.jar```. When I build the app and run it after `dist` it works. But I was not able to make it run with `sbt run`. The things I tried
These doesn't work:
* `sbt run` in **play app** with `addSbtPlugin("com.github.sbt" % "sbt-javaagent" % "0.1.8")` + settings (I know they work as they work in non play app)
* `sbt run` in **play app** with
``` 
run / fork := true,
run / forkOptions := ForkOptions().withRunJVMOptions(
   Vector(
     "-javaagent:opentelemetry-javaagent.jar"
))
```
These works:
* `sbt run` in **non play app** with `addSbtPlugin("com.github.sbt" % "sbt-javaagent" % "0.1.8")`
* `sbt runMain SomeHelloWorldMainClass` in **play app** with `addSbtPlugin("com.github.sbt" % "sbt-javaagent" % "0.1.8")`

What should I do?

**Comment by ihostage** [comment]
Hi, @fokot!
I'm not sure that it's possible for DEV-mode right now. For example Kamon wrote a [special Sbt plugin](https://github.com/kamon-io/sbt-kanela-runner/tree/master) to add their javaagent (Kanela) into Play DEV-mode.

**Comment by mkurz** [comment]
I _think_ `run / fork := true` does not even have an effect here. Play implements it's own run task and I think fork is not considered anyway.

So the app will be run in the same JVM process like sbt. So would you can try is:
1. Create a file called `.sbtopts` in the root of your project (the folder where you start `sbt`)
2. To that file add:
  ```
  -J-javaagent:opentelemetry-javaagent.jar
  ```

And restart sbt. Let me know if it works then.
Of course this is just for dev mode (when running sbt). For production you need to use another approach.

BTW: Instead of `.sbtopts` you can also use a `.jvmopts` file but without the `-J`:
```
-javaagent:opentelemetry-javaagent.jar
```

More details here: https://www.scala-sbt.org/1.x/docs/Command-Line-Reference.html#Sbt+runner+arguments

**Reply by ihostage** [reply]
Usually, tracing agents like as OTel agent instrument classes that loaded by PlatformClassloader in `preMain` function. That why I think it won't work with Play DEV-mode that loads classes by custom ClassLoaders and that why Kamon copy&paste our `run` task and add a few logic to instrument Play application in DEV-mode.

https://github.com/kamon-io/sbt-kanela-runner/blob/4af613be5dce2736cfefb90946bc684d1c3c8f25/sbt-kanela-runner-play-2.8/src/main/scala/kamon/instrumentation/sbt/play/KanelaReloader.scala#L279

But, of course, we can add an abstract solution for any Java agents (not only Kanela) in `Play-Run-Support` module in the future 🤔

**Reply by fokot** [reply]
I will try it out and let you know. Running in prod is no problem (it works) but I want to test things in dev mode. I do not need dynamic recompile. I just do not want to run `dist` all the time. Stopping and running run after code change would work for me. And it works with non play app.

**Reply by fokot** [reply]
With .sbtopts agent is starting, I see it logs, but I do not see any metrics, so probably it is not attached to `run`, even tough forking is off.

**Comment by fokot** [comment]
Is there a way to start play app with `runMain` and some class? I don't care about hot reload. I would restart sbt after every change

**Reply by ihostage** [reply]
Try to use `play.core.server.ProdServerStart` as a main class with `runMain` task.

**Reply by fokot** [reply]
I had some problems, but probably with our configuration

**Comment by fokot** [comment, marked as answer]
`runProd` is working with `sbt-javaagent` plugin.
